DataStructures/BinaryHeap.cs

```diff
@@ -0,0 +1,252 @@
+                T tmp = data[data.Count - 1];
+                data.RemoveAt(data.Count - 1);
+
+                if (idx < data.Count)
+                {
+                    if (comparer.Compare(tmp, data[idx]) > 0)
+                    {
+                        HeapifyDown(idx);
+                    }
+                    else
+                    {
+                        HeapifyUp(idx);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Swap the elements in the heap array at the given indices.
+        /// </summary>
+        /// <param name="idx1">First index.</param>
+        /// <param name="idx2">Second index.</param>
+        private void Swap(int idx1, int idx2)
+        {
+            T tmp = data[idx1];
+            data[idx1] = data[idx2];
+            data[idx2] = tmp;
+        }
+
+        /// <summary>
+        /// Recursive function to restore heap properties.
+        /// </summary>
+        /// <remarks>
+        /// Restores heap property by swapping the element at <paramref name="elemIdx"/>
+        /// with its parent if the element compares greater than its parent.
+        /// </remarks>
+        /// <param name="elemIdx">The element to check with its parent.</param>
+        private void HeapifyUp(int elemIdx)
+        {
+            var parent = (elemIdx - 1) / 2;
+
+            if (parent >= 0 && comparer.Compare(data[elemIdx], data[parent]) > 0)
+            {
+                Swap(elemIdx, parent);
+                HeapifyUp(parent);
+            }
+        }
+
+        /// <summary>
+        /// Recursive function to restore heap properties.
+        /// </summary>
+        /// <remarks>
+        /// Restores heap property by swapping the element at <paramref name="elemIdx"/>
+        /// with the larger of its children.
+        /// </remarks>
+        /// <param name="elemIdx">The element to check with its children.</param>
+        private void HeapifyDown(int elemIdx)
+        {
+            var left = (2 * elemIdx) + 1;
+            var right = (2 * elemIdx) + 2;
+
+            var leftLargerThanElem = left < Count && comparer.Compare(data[elemIdx], data[left]) < 0;
+            var rightLargerThanElem = right < Count && comparer.Compare(data[elemIdx], data[right]) < 0;
+            var leftLargerThanRight = left < Count && right < Count && comparer.Compare(data[left], data[right]) > 0;
+
+            if (leftLargerThanElem && leftLargerThanRight)
+            {
+                Swap(elemIdx, left);
+                HeapifyDown(left);
+            }
+            else if (rightLargerThanElem && !leftLargerThanRight)
+            {
+                Swap(elemIdx, right);
+                HeapifyDown(right);
+            }
+
+            // if left and right child values compare smaller than the current value, the heap property has been restored.
+            // no need to do anything else.
+            else
+            {
```

Comment:
Codacy found an issue: [Either remove or fill this block of code.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6375083)